Add Entry - Go: Loops (#592)
Codecademy/docs: content/go/concepts/loops/loops.md

---
Title: 'Loops'

> Description: 'There are several different types of loops in Go. All are written using the `for` command using different syntax.'

Reviewer: ```suggestion
Description: 'There are several different types of loops in Go. All are written using the `for` statement with different syntaxes.'
```

Subjects:
  - 'Code Foundations'
  - 'Computer Science'
Tags:
  - 'Control Flow'
  - 'Iterators'
  - 'Loops'
CatalogContent:
  - 'learn-go'
  - 'paths/back-end-engineer-career-path'
---

There are several different types of loops in Go. All are written using the `for` command using different syntax.

## Three Component Loop

The three component loop syntax is the same as ones used in C or Java. This is what is most often called a `for` loop in other languages.

### Syntax

```pseudo
for init; condition; post {
  statements
}
```

Here `init` is a statement initalizing an index variable. `condition` is a boolean expression checking the index variable before every loop, telling the loop to execute as long as it's true. And `post` is a statement executed after every loop to increment/decrement the index variable.

> ### Example

Reviewer: I think each example would benefit from having a brief explanation of what each does and the output for each.

Author: Done

```go
countdown := 10
for i := countdown; i > 0; i-- {
  fmt.Println(i) // Counts down from 10 to 1
}
```

## The While Loop

A "while" loop is one where the loop continues to execute as long as some condition remains true. In some languages this is implemented with a keyword `while`, in Go, the "while" loop is implemented with a `for` loop omitting the `init` and `post` statements.

### Syntax

```pseudo
for condition {
  statements
}
```

The loop will execute as long as `condition` remains true.

### Example

```go
countdown := 10
for countdown > 0 {
  fmt.Println(countdown) // counts down from 10 to 1
  countdown--
}
```

## An Infinite Loop

> If the condition statement is omitted along with the post and init statements, the loop will execute indefinitely unless a `break` statement is encountered.

Reviewer: Should post and init continue to be wrapped in backticks like it has been previously in this entry?
```suggestion
If the condition statement is omitted along with the `post` and `init` statements, the loop will execute indefinitely unless a `break` statement is encountered.
```

### Syntax

```pseudo
for {
  statements
}
```

### Example

```go
for {
  fmt.Println("Help! I'm trapped in a loop!") // executes forever
}
```

## A Range Loop

By using the `range` keyword, a `for` loop can step through the items in a collection such as a `map`, `slice`, `channel` or `string`.

### Syntax

```pseudo
for index, value = range collection {
  staements
}
```

Where `index` is an variable containing the index of the collection, `value` is a variable used to step through the values in `collection`, and `collection` is the collection the loop is stepping through.

### Example

```go

numbers := []string{"One","Two","Three"}
for i, n := range numbers {
  fmt.Println(i,n)
}

// Output:
// 0 One
// 1 Two
// 2 Three
```

## `break` and `continue`

The `break` and `continue` statements work in Go as they do in C and Java. A `break` statement halts execution of a loop and continues with the next statement after the loop. The `continue` statement skips execution to the next iteration of the loop.

### Examples

```go
for i := 0, i < 100, i++ {
  if i % 2 == 0 {
    continue
  }
  fmt.Println(i) // only prints odd numbers
}
```

> ```go

Reviewer: How do you feel about making this last example a codebyte?

Author: Good idea

A "while" loop runs as long as the condition is true. In some languages, this is implemented with a `while` keyword. In Go, however, the "while" loop is implemented with a `for` loop by omitting the `init` and `post` statements.

### Syntax

```pseudo
for condition {
  statements
}
```

The loop will execute as long as `condition` remains true.

### Example

This example behaves exactly the same way as the prior example, except the variable `countdown` is initalized outside the loop, and is decremented inside the loop itself. The condition still runs the loop as long as `countdown` is still greater than zero.

```go
countdown := 10
for countdown > 0 {
  fmt.Println(countdown) // counts down from 10 to 1
  countdown--
}
```

## An Infinite Loop

If the condition statement is omitted along with the post and init statements, the loop will execute indefinitely unless a `break` statement is encountered.

### Syntax

```pseudo
for {
  statements
}
```

### Example

This example will print the string "Help! I'm trapped in a loop!" until the program is halted externally.

```go
for {
  fmt.Println("Help! I'm trapped in a loop!") // executes forever
}
```

## A Range Loop

> By using the `range` keyword, a `for` loop can step through the items in a collection such as a array, map, slice, channel or string.

Reviewer 2: ```suggestion
By using the `range` keyword, a `for` loop can step through the items in a collection such as a array, map, slice, channel, or string.
```

### Syntax

```pseudo
for index, value = range collection {
  statements
}
```

Where `index` is a variable containing the index of the collection, `value` is a variable used to step through the values in `collection`, and `collection` is the collection the loop is stepping through.

### Example

> In the example below. a range loop steps through the elements of the slice `numbers`, printing the index-value pair for each element.

Reviewer 2: ```suggestion
In the example below, a range loop steps through the elements of the slice `numbers` and prints the index-value pair for each element.
```

---
Title: 'Loops'
Description: 'There are several different types of loops in Go. All are written using the `for` statement with different syntaxes.'
Subjects:
  - 'Code Foundations'
  - 'Computer Science'
Tags:
  - 'Control Flow'
  - 'Iterators'
  - 'Loops'
CatalogContent:
  - 'learn-go'
  - 'paths/back-end-engineer-career-path'
---

> There are several different types of loops in Go. All are written using the `for` statement with different syntaxes.

Reviewer 2: ```suggestion
There are several kinds of loops in Go, all written with a syntax similar to the `for` loop statement.
```

## Three-Component Loop

> The three-component loop syntax is the same as those used in C or Java. This is often called a `for` loop in other languages.

Reviewer 2: ```suggestion
In Go, loops are defined with a three-component syntax similar to what is used in `for` loops for other languages like C or Java.
```

Reviewer 2: Should something also be said about how Go does not use parenthesis for writing `for` loops?

### Syntax

```pseudo
for init; condition; post {
  statements
}
```

> Here `init` is a statement initalizing an index variable. `condition` is a boolean expression checking the index variable before every loop, telling the loop to execute as long as it's true. And `post` is a statement executed after every loop to increment/decrement the index variable.

Reviewer 2: ```suggestion
Here, the `init` statement initializes an index variable. Next, a `condition` is used to check the index variable against a boolean expression before every loop. As long as it's true, the iteration will go forward and `statements` inside the loop will execute. Lastly, the `post` statement executes after every loop, usually to increment/decrement the index variable.
```

```go
countdown := 10
for countdown > 0 {
  fmt.Println(countdown) // counts down from 10 to 1
  countdown--
}
```

## An Infinite Loop

If the condition statement is omitted along with the post and init statements, the loop will execute indefinitely unless a `break` statement is encountered.

### Syntax

```pseudo
for {
  statements
}
```

### Example

This example will print the string "Help! I'm trapped in a loop!" until the program is halted externally.

```go
for {
  fmt.Println("Help! I'm trapped in a loop!") // executes forever
}
```

## A Range Loop

By using the `range` keyword, a `for` loop can step through the items in a collection such as a array, map, slice, channel or string.

### Syntax

```pseudo
for index, value = range collection {
  statements
}
```

Where `index` is a variable containing the index of the collection, `value` is a variable used to step through the values in `collection`, and `collection` is the collection the loop is stepping through.

### Example

> This range loop will step through the elements of the slice `numbers`, printing the index and value of each element.

Reviewer 2: ```suggestion
In the example below. a range loop steps through the elements of the slice `numbers`, printing the index-value pair for each element.
```

```go

> 

Reviewer 2: ```suggestion
```

numbers := []string{"One","Two","Three"}
for i, n := range numbers {
  fmt.Println(i,n)
}

> 

Reviewer 2: I would personally end the code snippet here and write a sentence that leads into a separate output snippet.

// Output:
// 0 One
// 1 Two
// 2 Three
```

## `break` and `continue`

> The `break` and `continue` statements work in Go as they do in C and Java. A `break` statement halts execution of a loop and continues with the next statement after the loop. The `continue` statement skips execution to the next iteration of the loop.

Reviewer 2: ```suggestion
The `break` and `continue` statements work in Go as they do in C and Java.

* A `break` statement halts execution of a loop and continues with the next statement after the loop.
* The `continue` statement skips execution to the next iteration of the loop.
```

> ### Examples

Reviewer 2: ```suggestion
### Codebyte Examples
```

I think we should stick with just one Codebyte per entry. Especially since both of these examples produce a lot of `Println()` statements.

```codebyte/golang
package main
import "fmt"
for i := 0; i < 100; i++ {
  if i % 2 == 0 {
    continue
  }
  fmt.Println(i)
}
```

> 

Reviewer 2: In both of these Codebytes, the loop code needs to be nested inside of a `main() {...}` function in order to run:

```
package main

import "fmt"

func main() {
  i := 0
  for {
    i++
    if i == 50 {
      break
    }
    fmt.Println(i)
  }
}
```
